contrib/lib/src/databases.rs

```diff
@@ -0,0 +1,245 @@
+pub extern crate r2d2;
+
+use std::collections::BTreeMap;
+use std::fmt::{self, Display, Formatter};
+use std::marker::{Send, Sized};
+
+use rocket::config::{self, Value};
+
+pub use rocket_contrib_codegen::database;
+
+use self::r2d2::ManageConnection;
+
+#[cfg(any(feature = "diesel_sqlite_pool", feature = "diesel_postgres_pool", feature = "diesel_mysql_pool"))]
+pub extern crate diesel;
+
+#[cfg(feature = "postgres_pool")]
+pub extern crate postgres;
+#[cfg(feature = "postgres_pool")]
+pub extern crate r2d2_postgres;
+
+#[cfg(feature = "mysql_pool")]
+pub extern crate mysql;
+#[cfg(feature = "mysql_pool")]
+pub extern crate r2d2_mysql;
+
+#[cfg(feature = "sqlite_pool")]
+pub extern crate rusqlite;
+#[cfg(feature = "sqlite_pool")]
+pub extern crate r2d2_sqlite;
+
+#[cfg(feature = "cypher_pool")]
+pub extern crate rusted_cypher;
+#[cfg(feature = "cypher_pool")]
+pub extern crate r2d2_cypher;
+
+#[cfg(feature = "redis_pool")]
+pub extern crate redis;
+#[cfg(feature = "redis_pool")]
+pub extern crate r2d2_redis;
+
+/// A struct containing database configuration options.
+///
+/// This retrieves the key-value pairs specified in the `Rocket.toml` under
+/// the databases key for use in further configuring the desired connection.
+///
+/// Example configuration:
+/// ```toml
+/// [[global.databases]]
+/// name = "my_database"
+/// url = "postgres://postgres@localhost/my_database
+/// pool_size = 4
+/// ```
+pub struct DatabaseConfig<'a> {
+    /// The connection URL specified in the Rocket configuration
+    pub url: &'a str,
+    /// The size of the pool to be initialized. Defaults to the number of Rocket workers.
+    pub pool_size: i64,
+    /// Any extra options that are included in the configuration.
+    pub extras: &'a BTreeMap<String, Value>,
+}
+
+/// A custom error type to wrap multiple, disparate error types
+///
+/// This allows for returning the given error from the connection manager
+/// creation methods and the R2D2 connection pool methods.
+#[derive(Debug)]
+pub enum DbError<T: Display> {
+    Custom(T),
+    PoolError(r2d2::Error),
+}
+
+#[derive(Debug)]
+pub enum DatabaseConfigError {
+    MissingConfig,
+    MissingTable,
+    MissingConnectionString,
+    InvalidPoolSize,
+}
+
+impl DatabaseConfigError {
+    pub fn pretty_print(&self) {
+        error!("{}", self);
+    }
+}
+
+impl<'a> Display for DatabaseConfigError {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        match self {
```

Comment:
I find these error messages to be a bit confusing. When writing these, you want to put yourself in the position of a user who has received the error message and think about what would help you get from the error to the solution quickest. For instance, if I read "Could not find the specified connection configuration", I'd be confused about which "connection" this was talking about, where it tried to find it, and what I''m supposed to do.

The actual issue there seems to be that a configuration for a given database that the user has specified does not exist in the `databases` table in the active configuration; I would say something much closer to this.

One template that I've found useful for error message is to mention what the software thinks the user tried to do and where it failed to do that. You see this in compiler messages often, for instance: "Expected a value of type T here but found A instead."